File: python version/merge_same.py
import re
import os
import csv
import dataset

prefix_dict = {
    'Canon': ['IXUS', 'ELPH', 'IXY', 'Powershot', 'SX', 'EOS', 'SD', 'S', 'A', 'G'],
    'Casio': ['EX', 'QV'],
    'Fujifilm': ['Finepix', 'X', 'mini'],
    'GE': [],
    'Olympus': [r'SP', r'E', r'TG', r'SZ', r'Stylus', r'x', r'XZ', r'TG', r'SH'],
    'Panasonic': ['DMC'],
    'Samsung': ['EC'],
    'Sony': ['NEX', 'DSC', 'DMC', 'DSLR', 'SLT'],
    'Nikon': ['Nikon']
}

meaningful_postfix_dict = {
    'Fujifilm': r'(s|t|w)',
    'Sony': r'(r|s|(M3)|(II)|(III)|(IV))',
    'Nikon': r'(s|x|e|h)',
    'Canon': r'i'
}

auxiliary_same_dict = {
    'Panasonic': {('FT25', 'TZ25', 'ZS15'), ('FT4', 'TS4'), ('TZ18', 'ZS8'), ('TZ20', 'ZS10'),
                  ('TZ30', 'ZS20'), ('TZ40', 'ZS30'), ('TZ55', 'ZS25'), ('TZ7', 'ZS3'), ('TZ10', 'ZS7'),
                  ('TZ60', 'ZS40')},
    'Canon': {('IXUS 155', 'ELPH 150'), ('IXUS 510', 'ELPH 530'), ('IXUS 140', 'ELPH 130'),
              ('IXUS 145', 'ELPH 135'), ('IXUS 150', 'ELPH 140'), ('IXUS 240', 'ELPH 320'),
              ('IXUS 265', 'ELPH 340'), ('IXUS 1100', 'ELPH 510'), ('IXUS 310', 'ELPH 500'),
              ('IXUS 40', 'ELPH SD300'), ('IXUS 85', 'ELPH SD770'), ('IXUS 90', 'ELPH SD790'),
              ('IXUS 110', 'ELPH SD960'), ('IXUS 210', 'ELPH SD3500'), ('IXUS 700', 'ELPH SD500'),
              ('IXUS 800', 'ELPH SD700'), ('IXUS 850', 'ELPH SD800'), ('IXUS 900', 'SD900'),
              ('IXUS 950', 'ELPH SD850'), ('IXUS 960', 'SD960'), ('IXUS 970', 'SD890'),
              ('IXUS 980', 'SD990'), ('IXUS 1000', 'SD4500'), ('DS6041', '300D'), ('DS126191', '1000D'),
              ('X50', '1100D'), ('DS126311', '600D'), ('A3400', 'A34000'), ('IXUS i5', 'SD20'),
              ('IXUS i7', 'SD40'), ('EOS1', '1D')},
    'Fujifilm': {('JX520', 'JX500'), ('HS30', 'HS33'), ('HS25', 'HS28')},
    'Sony': {('RX100', 'RX100M3')}
}


def refine_model(brand_name, model_raw):
    model_clean = model_raw
    prefix_reg_list = prefix_dict[brand_name]
    if brand_name in meaningful_postfix_dict.keys():
        meaningful_postfix_reg = meaningful_postfix_dict[brand_name]
    else:
        meaningful_postfix_reg = None
    for reg in prefix_reg_list:
        if len(reg) == 1:
            prefix_ext = re.compile(r'^' + reg + r'[- ](.*)', re.I)

        else:
            prefix_ext = re.compile(r'^' + reg + r'[- ]?(.*)', re.I)
        if prefix_ext.findall(model_clean):
            tmp = prefix_ext.findall(model_clean)[0]
            if not re.match('[0-9]+[a-zA-Z]*$', tmp):
                model_clean = tmp
            else:
                model_clean = reg + tmp  # 去掉空格和-
    if re.match('[a-zA-Z]+[ -][0-9]+$', model_clean, re.I):
        tmp = re.findall('([a-zA-Z]+)[ -]([0-9]+)$', model_clean)
        model_clean = tmp[0][0] + tmp[0][1]
    if re.match('[a-zA-Z]+[0-9]+[a-zA-Z]+', model_clean):
        if not (meaningful_postfix_reg and re.match(r'.*' + meaningful_postfix_reg + r'$', model_raw, re.I)):
            model_clean = re.findall('([a-zA-Z]+[0-9]+)[-a-zA-Z]+', model_clean)[0]
    return model_clean.upper()



def merge(to_model, from_model, brand):
    dataset.model_index[brand][to_model] += dataset.model_index[brand][from_model]
    dataset.model_index[brand][to_model] = list(set(dataset.model_index[brand][to_model]))
    del dataset.model_index[brand][from_model]


def merge_same():

    for brand, brand_items in dataset.model_index.items():
        if brand not in prefix_dict.keys():
            continue
        model_dict = {}
        for model, model_items in brand_items.items():
            refined_model = refine_model(brand, model)
            if refined_model not in model_dict:
                model_dict[refined_model] = []
            model_dict[refined_model].append(model)

        # 合并打表的merg
        # filter_Kodak()
        # filter_Panasonic()e
        if brand in auxiliary_same_dict.keys():
            same_models = auxiliary_same_dict[brand]
            for same_model_tup in same_models:
                merge_list = []
                for item in same_model_tup:
                    tmp = refine_model(brand, item)
                    if tmp in model_dict.keys():
                        merge_list.append(tmp)
                    # print(item, refine_model(brand, item) in model_dict)
                # print(merge_list)
                if len(merge_list) > 1:
                    dest = merge_list[0]
                    for i in range(1, len(merge_list)):
                        model_dict[dest] = model_dict[dest] + model_dict[merge_list[i]]
                        print(model_dict[dest])
                        del model_dict[merge_list[i]]

        for item in model_dict.items():
            same_list = item[1]
            # 以最短的model名作为合并标准
            min_length = len(same_list[0])
            min_idx = 0
            for i in range(1, len(same_list)):
                if len(same_list[i]) < min_length:
                    min_length = len(same_list[i])
                    min_idx = i
            tmp = same_list[min_idx]
            same_list[min_idx] = same_list[0]
            same_list[0] = tmp
            if len(same_list) > 1:
                for i in range(1, len(same_list)):
                    print('Merge:', same_list[i], 'to', same_list[0])
                    merge(same_list[0], same_list[i], brand)

File: c++ version/merge_same.cpp
#include "merge_same.h"
#include "util.h"
#include "dataset.h"
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <list>
#include <regex>
#include <iostream>
#include <map>

using namespace std;


map<string, list<string>>prefix_dict = {
        {"Canon", {"IXUS", "ELPH", "IXY", "Powershot", "SX", "EOS", "SD", "S", "A", "G"}},
        {"Casio", {"EX", "QV"}},
        {"Fujifilm", {"Finepix", "X", "mini"}},
        {"Olympus", {"SP", "E", "TG", "SZ", "Stylus", "x", "XZ", "TG", "SH"}},
        {"Panasonic", {"DMC"}},
        {"Samsung", {"EC"}},
        {"Sony", {"NEX", "DSC", "DMC", "DSLR", "SLT", "ILCE"}},
        {"Nikon", {"Nikon"}}
};

map<string, string> meaningful_postfix_dict = {
        {"Fujifilm", "(s|t|w)"},
        {"Sony", "(r|s)"},
        {"Nikon", "(s|x|e|h)"},
        {"Canon", "i"}
};

map<string, list<list<string>>> auxiliary_same_dict = {
        {"Panasonic", {{"FT25", "TZ25", "ZS15"}, {"FT4", "TS4"}, {"TZ18", "ZS8"}, {"TZ20", "ZS10"}, {"TZ30", "ZS20"},
                       {"TZ40", "ZS30"}, {"TZ55", "ZS25"}, {"TZ7", "ZS3"}, {"TZ10", "ZS7"}, {"TZ60", "ZS40"}}},
        {"Canon", {{"IXUS 155", "ELPH 150"}, {"IXUS 510", "ELPH 530"}, {"IXUS 140", "ELPH 130"},
                   {"IXUS 145", "ELPH 135"}, {"IXUS 150", "ELPH 140"}, {"IXUS 240", "ELPH 320"},
                   {"IXUS 40", "SD300"}, {"IXUS 85", "SD770"}, {"IXUS 90", "SD790"},
                   {"IXUS 110", "SD960"}, {"IXUS 210", "SD3500"}, {"IXUS 700", "SD500"},
                   {"IXUS 800", "SD700"}, {"IXUS 850", "SD800"}, {"IXUS 900", "SD900"},
                   {"IXUS 950", "SD850"}, {"IXUS 960", "SD960"}, {"IXUS 970", "SD890"},
                   {"IXUS 980", "SD990"}, {"IXUS 1000", "SD4500"}, {"DS6041", "300D"}, {"DS126191", "1000D"},
                   {"X50", "1100D"}, {"DS126311", "600D"}, {"A3400", "A34000"}, {"IXUS i5", "SD20"},
                   {"IXUS i7", "SD40"}, {"1D", "EOS1"}, {"5D", "EOS5"}, {"7D", "EOS7"}}},
        {"Fujifilm", {{"JX520", "JX500"}, {"HS30", "HS33"}, {"HS25", "HS28"}}},
{"Olympus", {{"EM5", "M5"}}},
        {"Sony", {{"RX100", "RX100M3", "RX100M2", "RX100M3B"}}}
};



string refine_model(const string& brand_name, const string& model_raw) {
    string model_clean = model_raw;
    const list<string>& prefixes = prefix_dict[brand_name];
    string meaningful_postfix_reg;
    if (meaningful_postfix_dict.find(brand_name) != meaningful_postfix_dict.end()) {
        meaningful_postfix_reg = meaningful_postfix_dict[brand_name];
    }
    for (const string& reg: prefixes) { // 前缀标准化
        string reg_ext_str;
        if (reg.size() == 1) {
            reg_ext_str = "^" + reg + "[- ](.*)";
        }
        else {
            reg_ext_str = "^" + reg + "[- ]?(.*)";
        }
        regex pattern_1(reg_ext_str, regex_constants::icase);
        smatch results_1;
        bool success = regex_match(model_clean, results_1, pattern_1);
        if (success) {
            string tmp_str = *(results_1.begin()+1);
            if (!regex_match(tmp_str, regex("[0-9]+[a-zA-Z]*$", regex_constants::icase))) {
                model_clean = tmp_str;
            }
            else {
                model_clean = reg + tmp_str;
            }
        }
    }

    regex pattern_2("([a-zA-Z]+)[ -]([0-9]+)$", regex_constants::icase);
    smatch results_2;
    bool success = regex_match(model_clean, results_2, pattern_2);
    if (success) {
        string tmp_1 = *(results_2.begin()+1);
        string tmp_2 = *(results_2.begin()+2);
        model_clean = tmp_1 + tmp_2;
    }

    string post_reg_str = ".*" + meaningful_postfix_reg + "$";
    if (regex_match(model_clean, regex("[a-zA-Z]+[0-9]+[a-zA-Z]+", regex_constants::icase))) {
        if (!(!meaningful_postfix_reg.empty() && regex_match(model_raw, regex(post_reg_str, regex_constants::icase)))) {
            regex pattern_3 ("([a-zA-Z]+[0-9]+)[-a-zA-Z]+");
            smatch result;
            regex_search(model_clean, result, pattern_3);
            model_clean = *(result.begin()+1);
        }
    }
    return str_upper(model_clean);

}



inline void merge(const string& to_model, const string& from_model, const string& brand) {
    if (from_model == to_model) return;
    model_index[brand][to_model].insert(model_index[brand][from_model].begin(),
                                        model_index[brand][from_model].end());
    model_index[brand].erase(from_model);
}



void merge_same() {
    for (const auto& brand_item: model_index) {
        string brand_name = brand_item.first;
        if (prefix_dict.find(brand_name) == prefix_dict.end()) {
            continue;
        }
        cout << "Merge same: " << brand_name << endl;
        map<string, list<string>> model_dict;
        for (const auto& model_item: brand_item.second) {
            string model_name = model_item.first;
            string refined_model = refine_model(brand_name, model_name);
            //cout << "REFINE: " << model_name << "->" << refined_model << endl;
            model_dict[refined_model].push_back(model_name);
        }


//        for (auto item: model_dict) {
//            if (item.second.size() >= 2) {
//                for (string i: item.second) {
//                    cout << i << ",";
//                }
//                cout << endl;
//            }
//        }


        //unordered_map<string, unordered_set<string>>
        if (auxiliary_same_dict.find(brand_name) != auxiliary_same_dict.end()) {
            list<list<string>> same_models = auxiliary_same_dict[brand_name];
            for (const list<string>& same_list: same_models) {
                vector<string> merge_list;
                for (const string& item: same_list) {
                    string tmp = refine_model(brand_name, item);
                    if (model_dict.find(tmp) != model_dict.end()) {
                        merge_list.push_back(tmp);
                    }
                }

                if (merge_list.size() > 1) {

                    cout << "merge list: ";
                    for (string s: merge_list) {
                        cout << s << " ";
                    }
                    cout << endl;



                    string dest_model = merge_list[0];
                    for (int i = 1; i < merge_list.size(); i++) {
			if(merge_list[i]==dest_model) continue;
                        for (string s: model_dict[merge_list[i]]) {
                            model_dict[dest_model].push_back(s);
                        }
//                        model_dict[dest_model].insert(model_dict[merge_list[i]].begin(),
//                                                      model_dict[merge_list[i]].end());
                        cout << "Model dict: " << merge_list[i] << " to " << dest_model << endl;
                        model_dict.erase(merge_list[i]);
                    }
                }
            }
        }

        /*
        cout << "model dict:" << brand_name << endl;
        for (auto item: model_dict) {
            cout << item.first << ":  ";
            for (string x: item.second) {
                cout << x << ";";
            }
            cout << endl;
        }*/




        //unordered_map<string, unordered_set<string>>
        for (auto item: model_dict) {
            vector<string> same_list {item.second.begin(), item.second.end()};
            sort(same_list.begin(), same_list.end());

//            int min_length = same_list[0].length();
//            int min_index = 0;
//            for (int i = 1; i < same_list.size(); i ++) {
//                if (same_list[i].length() < min_length) {
//                    min_length = same_list[i].length();
//                    min_index = i;
//                }
//            }
//            string tmp = same_list[min_index];
//            same_list[min_index] = same_list[0];
//            same_list[0] = tmp;


            //if (same_list.size() > 1) {
            for (int i = 0; i < same_list.size(); i++) {
                //cout << "Merge: " << same_list[i] <<  " to " << item.first << endl;
                merge(item.first, same_list[i], brand_name);
//                    if (item.first != same_list[i]) {
//                        model_index[brand_name].erase(item.first);
//                    }
            }
            //}
        }
    }
}
